### ISSUE TITLE
IO::Handle::setvbuf not defined properly

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6539]
Migrated from [rt.perl.org#22339](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=22339) (status was 'resolved')

Searchable as RT22339$

**Comment by p5pRT** [comment]
### From @mjdominus

#### Created by @mjdominus



Even though my platform certainly does suport 'setvbuf'\, IO&#8203;::Handle
refuses to do it&#8203;:

&nbsp;        use Config;
&nbsp;        use IO&#8203;::Handle;

&nbsp;        print "d\_setvbuf is '$Config\{d\_setvbuf\}'\\n";
&nbsp;        print "\_IONBF is "\, IO&#8203;::Handle&#8203;::\_IONBF\, "\\n";

&nbsp;        print "But\.\.\.\\n";

&nbsp;        STDOUT\->setvbuf\(undef\, IO&#8203;::Handle&#8203;::\_IONBF\, 0\);

The output&#8203;:

&nbsp;        d\_setvbuf is 'define'
&nbsp;        \_IONBF is 2
&nbsp;        But\.\.\.
&nbsp;        IO&#8203;::Handle&#8203;::setvbuf not implemented on this architecture at /tmp/setvbuf line 9\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.0:

Configured by mjd at Thu Apr 17 11:57:37 EDT 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.2-2, archname=i586-linux
    uname='linux plover.com 2.4.2-2 #1 sun apr 8 19:37:14 edt 2001 i586 unknown '
    config_args='-des'
    hint=previous, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -I/usr/local/include -I/usr/include/gdbm -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -I/usr/include/gdbm'
    ccversion='', gccversion='2.96 20000731 (Red Hat Linux 7.1 2.96-81)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldl -lm -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lcrypt -lutil
    libc=/lib/libc-2.2.2.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.2.4'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /usr/local/lib/perl5/5.8.0/i586-linux
    /usr/local/lib/perl5/5.8.0
    /usr/local/lib/perl5/site_perl/5.8.0/i586-linux
    /usr/local/lib/perl5/site_perl/5.8.0
    /usr/local/lib/perl5/site_perl/5.7.3
    /usr/local/lib/perl5/site_perl/5.7.2
    /usr/local/lib/perl5/site_perl/5.6.1
    /usr/local/lib/perl5/site_perl/5.6.0
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.8.0:
    HOME=/home/mjd
    LANG=C
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/lib:/usr/lib:/usr/X11R6/lib
    LOGDIR (unset)
    PATH=/home/mjd/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/usr/games:/sbin:/usr/sbin:/usr/local/bin/X11R6:/usr/local/bin/mh:/data/mysql/bin:/usr/local/bin/pbm:/usr/local/bin/ezmlm:/home/mjd/TPI/bin:/usr/local/teTeX/bin:/usr/local/mysql/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

Mark\-Jason Dominus \(via RT\) wrote&#8203;:
> Even though my platform certainly does suport 'setvbuf'\, IO&#8203;::Handle
> refuses to do it&#8203;:
\[\.\.\.\]
> 
>         d\_setvbuf is 'define'
>         \_IONBF is 2
>         But\.\.\.
>         IO&#8203;::Handle&#8203;::setvbuf not implemented on this architecture at /tmp/setvbuf line 9\.

\[\.\.\.\]
>     useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef

setbuf/setvbuf aren't available with a PerlIO\-enabled perl\.
I suggest to simply document this\.

**Comment by p5pRT** [comment]
### From @mjdominus

Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr>&#8203;:
> setbuf/setvbuf aren't available with a PerlIO\-enabled perl\.

Thanks\.

Here's my problem\.  I want two processes to access the same \(small\)
file concurrently\.  Process A reads part of the file\.  Process B
modifies part of the file\.  Then process A wants to see the change\.
But it can't\, because the block it just read is buffered\, and
subsequent reads are all coming from the buffer\, which records the
state of the file before the change\, instead of from the real file\.

I thought my choices were

a\) use setvbuf\(\) to discard the buffer\, perhaps temporarily\, or

b\) discard the buffered I/O system entirely and rewrite everything in
terms of sysread\(\)\, effectively reimplementing the entire buffered I/O
subsystem in Perl\.

But now it appears that my only choice is \(b\)\.  This isn't very
attractive\.  Is there some other option I've overlooked?

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, May 28\, 2003 at 08&#8203;:58&#8203;:21AM \-0400\, Mark Jason Dominus wrote&#8203;:
> Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr>&#8203;:
> > setbuf/setvbuf aren't available with a PerlIO\-enabled perl\.
> 
> Thanks\.
> 
> Here's my problem\.  I want two processes to access the same \(small\)
> file concurrently\.  Process A reads part of the file\.  Process B
> modifies part of the file\.  Then process A wants to see the change\.
> But it can't\, because the block it just read is buffered\, and
> subsequent reads are all coming from the buffer\, which records the
> state of the file before the change\, instead of from the real file\.
> 
> I thought my choices were
> 
> a\) use setvbuf\(\) to discard the buffer\, perhaps temporarily\, or
> 
> b\) discard the buffered I/O system entirely and rewrite everything in
> terms of sysread\(\)\, effectively reimplementing the entire buffered I/O
> subsystem in Perl\.
> 
> But now it appears that my only choice is \(b\)\.  This isn't very
> attractive\.  Is there some other option I've overlooked?

My understanding is that one never was allowed to call setvbuf\(\) midway on a
stream&#8203;:

C99 \(spit\, but all I have access to\) section 7\.19\.5\.6 paragraph 2&#8203;:

2    The setvbuf function may be used only after the stream pointed to by
&nbsp;     stream has been associated with an open file and before any other
&nbsp;     operation \(other than an unsuccessful call to setvbuf\) is performed on
&nbsp;     the stream\.

The fact that it always seems to work is undefined behaviour\.
I only know this as a side effect of this thread&#8203;:

http&#8203;://groups\.google\.com/groups?hl=en&lr=&ie=UTF\-8&th=6a3278c9c3b3534e&rnum=1 

\(the code that generated the test output called setvbuf\(\) on an already
used stream\, and someone politely pointed out my error\. I thought that it
was in public\, but I can't find it in the thread\)

Nicholas Clark

**Comment by p5pRT** [comment]
### From kstar@cpan.org

On May 28\, Mark\-Jason Dominus wrote&#8203;:
> Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr>&#8203;:
> > setbuf/setvbuf aren't available with a PerlIO\-enabled perl\.
> 
> Thanks\.
> 
> Here's my problem\.  I want two processes to access the same \(small\)
> file concurrently\.  Process A reads part of the file\.  Process B
> modifies part of the file\.  Then process A wants to see the change\.
> But it can't\, because the block it just read is buffered\, and
> subsequent reads are all coming from the buffer\, which records the
> state of the file before the change\, instead of from the real file\.
> 
> I thought my choices were
> 
> a\) use setvbuf\(\) to discard the buffer\, perhaps temporarily\, or
> 
> b\) discard the buffered I/O system entirely and rewrite everything in
> terms of sysread\(\)\, effectively reimplementing the entire buffered I/O
> subsystem in Perl\.
> 
> But now it appears that my only choice is \(b\)\.  This isn't very
> attractive\.  Is there some other option I've overlooked?

&nbsp;    How about mmap\(\)\, or closing and re\-opening the read\(\) filehandle?

&nbsp;    \- Kurt

**Comment by p5pRT** [comment]
### From @gbarr

On Wed\, May 28\, 2003 at 08&#8203;:58&#8203;:21AM \-0400\, Mark Jason Dominus wrote&#8203;:
> Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr>&#8203;:
> > setbuf/setvbuf aren't available with a PerlIO\-enabled perl\.
> 
> Thanks\.
> 
> Here's my problem\.  I want two processes to access the same \(small\)
> file concurrently\.  Process A reads part of the file\.  Process B
> modifies part of the file\.  Then process A wants to see the change\.
> But it can't\, because the block it just read is buffered\, and
> subsequent reads are all coming from the buffer\, which records the
> state of the file before the change\, instead of from the real file\.
> 
> I thought my choices were
> 
> a\) use setvbuf\(\) to discard the buffer\, perhaps temporarily\, or
> 
> b\) discard the buffered I/O system entirely and rewrite everything in
> terms of sysread\(\)\, effectively reimplementing the entire buffered I/O
> subsystem in Perl\.
> 
> But now it appears that my only choice is \(b\)\.  This isn't very
> attractive\.  Is there some other option I've overlooked?

You could try using seek\.

 my $pos = tell\($fh\);
 seek\($fh\,0\,0\);
 seek\($fh\,$pos\,0\);

Graham\.

**Comment by p5pRT** [comment]
### From @mjdominus


Nicholas Clark \<nick@&#8203;ccl4\.org>&#8203;:
> > a\) use setvbuf\(\) to discard the buffer\, perhaps temporarily\, or
> 
> My understanding is that one never was allowed to call setvbuf\(\) midway on a
> stream&#8203;:

Right\, thanks\.  Please forget about the 'perhaps temporarily' part\.  I
added this as an afterthought in my message anyway\.  The essential
problem is that I would like to do

&nbsp;        $rec = \<FH>;

but that implies some sort of buffering\, and the buffering confounds
concurrent access\.  So I planned to use setvbuf to disabled the
buffering while stil permitting use of \<FH>\.


Kurt Starsinic \<kstar@&#8203;cpan\.org>&#8203;:
>    How about mmap\(\)\, 

I need features that are provided by standard perl\, preferably back to
5\.005\.  As far as I know there's no entry to mmap except through
nonstandard CPAN modules\.  PerlIO uses it internally\, if it's
available\, but doesn't make it available to me\.

Kurt Starsinic \<kstar@&#8203;cpan\.org>&#8203;:
> or closing and re\-opening the read\(\) filehandle?

Not an option\, I'm afraid\.  It may not be re\-openable\.  

Using

&nbsp;        open $newfh\, "\<&="\.fileno\($oldfh\) 

seems not to work for some reason that I haven't investigated yet\.

Graham Barr \<gbarr@&#8203;pobox\.com>&#8203;:
> You could try using seek\.

That works on my platform\, but I worry that it's not portable\.  I
can't think of any reason why 'seek' should necessarily discard the
buffered data\.  Perhaps I'm overlooking something?

**Comment by p5pRT** [comment]
### From @jhi

For what's it worth\, I now documented \(change \#19631\) in IO&#8203;::Handle that
the IO&#8203;::Handle&#8203;::setvbuf is unlikely to available in Perls 5\.8\.0 and later\.
I thought of changing the error message too\, but then backpedaled since
"platform" also IHMO includes how Perl has been configured\, for example
whether PerlIO is in use\.

**Comment by p5pRT** [comment]
@jhi - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From nick.ing-simmons@elixent.com

Kurt Starsinic \<kstar@&#8203;cpan\.org> writes&#8203;:
>On May 28\, Mark\-Jason Dominus wrote&#8203;:
>> Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;free\.fr>&#8203;:
>> > setbuf/setvbuf aren't available with a PerlIO\-enabled perl\.
>> 
>> Thanks\.
>> 
>> Here's my problem\.  I want two processes to access the same \(small\)
>> file concurrently\.  Process A reads part of the file\.  Process B
>> modifies part of the file\.  Then process A wants to see the change\.
>> But it can't\, because the block it just read is buffered\, and
>> subsequent reads are all coming from the buffer\, which records the
>> state of the file before the change\, instead of from the real file\.
>> 
>> I thought my choices were
>> 
>> a\) use setvbuf\(\) to discard the buffer\, perhaps temporarily\, or

As Nick C has said\, as far as I know once you have called setvbuf\(\)
it is done and all future reads are unbuffered \(if that is what you set\)\.

>> 
>> b\) discard the buffered I/O system entirely and rewrite everything in
>> terms of sysread\(\)\, effectively reimplementing the entire buffered I/O
>> subsystem in Perl\.
>> 
>> But now it appears that my only choice is \(b\)\.  This isn't very
>> attractive\.  Is there some other option I've overlooked?

seek\(FH\,0\,1\);  \# re\-position here

Will discard the buffer contents on a read stream\.

>
>    How about mmap\(\)\, or closing and re\-opening the read\(\) filehandle?

I wonder if :mmap layer has finaly found a use?


>
>    \- Kurt
\-\- 
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/

**Comment by p5pRT** [comment]
### From nick.ing-simmons@elixent.com

Mark Jason Dominus \<mjd@&#8203;plover\.com> writes&#8203;:
>You said&#8203;:
>> 
>> seek\(FH\,0\,1\);  \# re\-position here
>> 
>> Will discard the buffer contents on a read stream\.
>
>Is that guaranteed to work on all platforms\, all implementations\,

Not yet obviously or we would not have this thread\.

But is guaranteed to work on a :unix&#8203;:perlio stream\.

>forever?  I don't expect so\.  The 'correct' solution would seem to be
>to use 'setvbuf'\, which is part of the C standard\.

But PerlIO is not a C standard\. 

There may be no stdio involved at all\.

Personally I refuse to adopt a setvbuf\(\) style API which packs too 
many options into its args \- including some which expose addresses
of buffers which is not perl friendly\.


\-\- 
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/

**Comment by p5pRT** [comment]
### From nick.ing-simmons@elixent.com

Mark Jason Dominus \<mjd@&#8203;plover\.com> writes&#8203;:
>> Mark Jason Dominus \<mjd@&#8203;plover\.com> writes&#8203;:
>> >You said&#8203;:
>> >> 
>> >> seek\(FH\,0\,1\);  \# re\-position here
>> >> 
>> >> Will discard the buffer contents on a read stream\.
>> >
>> >Is that guaranteed to work on all platforms\, all implementations\,
>> 
>> Not yet obviously or we would not have this thread\.
>
>"Not yet" would seem to be misleading\.   It can never be guaranteed
>to work everywhere

PerlIO\_seek\(\) or pp\_seek\(\) can be defined to make it true everywhere\.
However I would wish we had a 'flush' keyword so that 

&nbsp;   $fh\->flush 

would do the right thing

\[If you suggested IO&#8203;::Handle&#8203;::flush\(\) be defined to do the right thing
for input streams I would support that\.\]

>
>> But is guaranteed to work on a :unix&#8203;:perlio stream\.
>
>That seems to me like a strange decision\.  

Hmm\, you may be right there \- I have forgotten the original cause 
for the problem\.

>seek\(FH\,0\,1\) is obviously a
>no\-op\, and therefore a candidate for optimization\.  The reason
>seek\(FH\,0\,1\) worries me so much is that I can imagine reasonable
>implementations that ignore it\.  For example\, if stdio or perlio were
>simply to read buffers on demand\, postponing the flush until a reading
>operation occurred outside the current buffer\, seek\(FH\,0\,1\) would have
>no effect\, as would
>
>        seek\(FH\,$N\,1\); seek\(FH\,\-$N\,1\);
>
>By promising that seek\(FH\,0\,1\) will flush the buffer\, you are
>foreclosing the possibility of this design and of adding which might
>be an obvious optimization\.  

I agree with most of that in principle \- but the optimization is just 
as obvious to the application code\, and nop\-seek is a thing that 
has meaning for various reasons\. e\.g\. stdio stuff has already lead to 
various nop\-seeks to switch between read/write on same stream\. 
There is also code with such nop\-seeks to get the stdio stream and 
sysread\(\) in sync\. \.\.\.


>
>The Single Unix Specification does not specify that fseek\(\) will
>discard the buffered data on an input stream\, presumably for some
>reason similar to this one\.

But does say that to switch between read/write on a RW stream you 
must seek\.

>
>But if you are sure that is how you want to have it done\, that is fine
>with me\.  

What \_I\_ want you to do is specify in a PerlIO way that you want 
no buffering\. If the defaults don't suit \- then say what does&#8203;:

open\(my $fh\,&#8203;:"\<&#8203;:unix"\, \# no buffer layer here
&nbsp;     

>Is it in the documentation?  If not\, would you be willing to
>put it there?
>
>See below for an alternative\.
>> 
>> Personally I refuse to adopt a setvbuf\(\) style API which packs too 
>> many options into its args \- including some which expose addresses
>> of buffers which is not perl friendly\.
>
>I understand about the buffers\.  But all I want is \_IONBF\, which\,
>unlike the other options\, is very simple&#8203;: no buffering at all\.  I am
>not asking for PerlIO to support line buffering or anything like that\.
>
>The current implementation replaces setvbuf with
>
>        IO&#8203;::Handle&#8203;::setvbuf not implemented on this architecture
>
>Would you be willing to support \_IONBF\, and leave all other uses of 
>setvbuf as "not implemented on this architecture"?


My personal goal is never to have to use IO&#8203;::\* for anything\.
\[So if someone patched IO\.xs to do what you suggest here I don't care\.
 Whoever does that needs to reconcile what \_IONBF means if asked
 for on when layer needs buffering to work e\.g\. 
 crlf/quoted\_printable/gzip/encoding etc\.\]
&nbsp;    
Why should I need to 

use IO&#8203;::Handle qw\(\_IONBF\);

&nbsp;  $fh\->setvbuf\(undef\,\_IONBF\(\)\,undef\);

When what I really want is 

&nbsp;     $fh\->unbuffered;

If if I wanted it that way then why did I open it buffered in the 1st place?

And in particular if want it buffered but occasionaly 
want to discard any pre\-read input that sould be easy too\.


\-\- 
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/

**Comment by p5pRT** [comment]
### From @chipdude

According to Mark\-Jason Dominus&#8203;:
>   $rec = \<FH>;
> but that implies some sort of buffering\, and the buffering confounds
> concurrent access\.

You could resort to sysread/syswrite/sysseek and just Not Do That
\(wrt readline\)\.
\-\- 
Chip Salzenberg               \- a\.k\.a\. \-               \<chip@&#8203;pobox\.com>
"I wanted to play hopscotch with the impenetrable mystery of existence\,
&nbsp;    but he stepped in a wormhole and had to go in early\."  // MST3K

**Comment by p5pRT** [comment]
### From @mjdominus

Chip Salzenberg \<chip@&#8203;pobox\.com>&#8203;:
> According to Mark\-Jason Dominus&#8203;:
> >   $rec = \<FH>;
> > but that implies some sort of buffering\, and the buffering confounds
> > concurrent access\.
> 
> You could resort to sysread/syswrite/sysseek and just Not Do That
> \(wrt readline\)\.


Back on 28 May\, I had said&#8203;:

>> I thought my choices were
>> 
>> a\) use setvbuf\(\) to discard the buffer\, perhaps temporarily\, or
>> 
>> b\) discard the buffered I/O system entirely and rewrite everything in
>> terms of sysread\(\)\, effectively reimplementing the entire buffered I/O
>> subsystem in Perl\.
>> 
>> But now it appears that my only choice is \(b\)\.  This isn't very
>> attractive\.  Is there some other option I've overlooked?
>> 

What you appear to be suggesting is \(b\)\.  Have I missed something?

**Comment by p5pRT** [comment]
### From @chipdude

According to Mark\-Jason Dominus&#8203;:
> >> b\) discard the buffered I/O system entirely and rewrite everything in
> >> terms of sysread\(\)\, effectively reimplementing the entire buffered I/O
> >> subsystem in Perl\.
> 
> What you appear to be suggesting is \(b\)\.  Have I missed something?

No\, I missed Careful Reading in Politeness Class\.  \-\-me

"Never mind\."
\-\- 
Chip Salzenberg               \- a\.k\.a\. \-               \<chip@&#8203;pobox\.com>
"I wanted to play hopscotch with the impenetrable mystery of existence\,
&nbsp;    but he stepped in a wormhole and had to go in early\."  // MST3K

**Comment by p5pRT** [comment]
### From @mjdominus

Nick Ing\-Simmons \<nick\.ing\-simmons@&#8203;elixent\.com>&#8203;:
> What \_I\_ want you to do is specify in a PerlIO way that you want 
> no buffering\.

Actually what I really want is a way to discard the input buffer
whenever I want\, and also a way to detect whether this will be
possible\, given only the handle\.  If you say that seek\(H\,0\,1\) will
always work on all handles in all PerlIO implementations\, that is good
enough for me\.

Thanks\.

**Comment by p5pRT** [comment]
### From @jhi

I believe this got resolved&#8203;: for PerlIO there's a workaround way\, and I documented
that IO&#8203;::Handle&#8203;::setvbuf is not implemented under PerlIO\.  I'm marking the problem
ticket as resolved\.

**Comment by p5pRT** [comment]
@jhi - Status changed from 'open' to 'resolved'
